cirq/linalg/diagonalize.py

```diff
@@ -218,15 +218,15 @@ def bidiagonalize_real_matrix_pair_with_symmetric_products(
     # Merge the fixup factors into the initial diagonalization.
     left_adjust = combinators.block_diag(overlap_adjust, extra_left_adjust)
     right_adjust = combinators.block_diag(overlap_adjust.T, extra_right_adjust)
-    left = left_adjust.T.dot(base_left.T)
-    right = base_right.T.dot(right_adjust.T)
+    left = cast(np.ndarray, left_adjust.T.dot(base_left.T))
+    right = cast(np.ndarray, base_right.T.dot(right_adjust.T))
```

Comment:
So only the cast on `left` is necessary and I'm trying to figure out why ......

Reply:
it seems like in the general case

```python
x: np.ndarray
y: np.ndarray
x.dot(y) --> Union[number[Any], ndarray]
```

but 

```python
np.dot(x, y) --> ndarray
```

weird

Reply:
the former is defined in the top-level `__init__.pyi` type stub file. The latter may be due to the docstring of the skeleton implementation in numpy/core/multiarray.py (that dispatches to c ufuncs). Maybe an oversight on numpy's part? I still don't know why `right` doesn't need a cast in the above.

Reply:
I think the former is the correct signature but the latter is the easier signature. I've re-written this function to use `cirq.dot` and `np.dot` to keep everything as `ndarray`s